### PR TITLE
Ghaf user accounts

### PR DESCRIPTION
**Pull request by mbssrc**
The changes address the separation between declarative and runtime users.Non-declarative users are introduced that allow changes without re-building the configuration, and are purely device specific.

For more consistent declarative user management, 'userborn' is introduced, which increases consistency by managing user parameter changes across re-builds. It is enabled for host and all VMs by default. The NixOS configuration `users.mutableUsers` is set to false, thus the generic user tools cannot be used at runtime (`passwd`, `useradd`, etc.). Further, a new user configuration file allows to conveniently override and set user attributes across the system (e.g., password updates).

To manage non-declarative users, `systemd-homed` is used. It offers a variety of concepts that are in line with our requirements such as on-the-fly user creation, encryption, CIFS-integration, and more. For the login-user, password
change with `passwd` should work.

## Code structure
```
common.nix - common settings
admin.nix - admin account settings
desktop.nix - user accounts for desktop setup
managed.nix - declarative users
reference/personalize/accounts.nix - entry point for user/password management
```
### Admin account
The 'ghaf' user account is now the admin account (+wheel), and enabled by default in host and all VMs. As this account is for administrative purposes, it should not run a desktop session.

### Login/desktop user account
The 'loginUser' account can be enabled and sets up a non-declarative user with a reserved UID. It is a self-contained account, bootstrapped via the home folder, that currently runs the user desktop session. Files in `/var/lib/systemd/home` bind the home folder to the machine.

### Auxiliary accounts
Two auxiliary accounts are available that share the login users UID to
keep these UIDs consistent across VMs:
  1) Proxy user
  This user is used in system VMs that provide services. These services
  are currently accessible via the dbus proxy, and require the same UID.
  2) App user
  This user is used in app VMs to run the user sessions (including applications).
  The shared UID is helpful to map access rights across machines and support
  legacy functionality.

### Other (declarative/managed) user accounts
While any additional user accounts may be freely created and administered, a template for configuration managed users is provided.

## Future work
This patch introduces new account management on Ghaf. Based on the changes, future work is required to extend it.

### Extending login-user functionality
Currently, a minimalistic setup script runs on first GUI-VM boot. The login
user setup may be extended with:

  - Improvement of user creation script
  - Graphical interface for user creation

  - Feature integration: (supported by homed)
    - CIFS/remote user storage integration
    - FIDO token integration
    - User ssh keys
    - Avatar/Background/Locale/Timezone/Location
    - External home support (e.g., USB or network storage)

  - Potential re-work to run graphical session as static user
  - Potential multi-user system with migratable data

### Further user account improvements

- removing (hardcoded) ssh dependencies
- removing ssh root access, e.g., update user (PoC available)
- password policies, considering declarative users passwords are in nix store

### User data implications
While the login user setup provides some containment, currently user data is still
spread across the system, such as
  - Persistent user-related platform data (e.g., wifi passwords)
  - Persistent user data in app VMs (e.g., browser profiles/passwords)
  - File sharing between VMs

Respective mechanisms are currently under investigation.

## Implementation notes

  - VMs now run with stage1 systemd
  - VM storage shares have been re-named for consistency (<name>-vm)
  - impermanence flake input pinned to userborn patch
  - /etc/machine-id (gui-vm) is currently hardcoded as login user identity file depends on it.
    It should be generated on first boot and persistet. Workaround is available upstream (after
    userborn patch) in impermanence but does not seem to work with our setup, investigation required
  - reverts .face patch. There is an avatar setting, but afaik file needs to be present during setup
  - known login inconveniences: wrong password entry on lock-screen results in multiple errors, user needs to click
    login if fprint is enabled (only for new login, not lock)
  - security warning about systemd-generators being writable
  - VM shares have been renamed - if not flashed all previous user data is inaccessible

## Checklist for things done

- [x] Summary of the proposed changes in the PR description
- [x] More detailed description in the commit message(s)
- [x] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [x] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [ ] PR linked to architecture documentation and requirement(s) (ticket id)
- [x] Test procedure described (or includes tests). Select one or more:
  - [x] Tested on Lenovo X1 `x86_64`
  - [ ] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [x] Author has run `make-checks` and it passes
- [x] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [x] Author has added reviewers and removed PR draft status
- [x] Change requires full re-installation
- [ ] Change can be updated with `nixos-rebuild ... switch`

## Instructions for Testing

As this PR affects all of Ghaf, extensive testing is required. All desktop functionality should
be tested, and all targets verified. All VMs are running with systemd stage1 boot, and 
all cross-vm functionality that previously used ssh + ghaf user is affected.

As this patch also re-names VM shares, target should be flashed before testing.

- [x] List all targets that this applies to: Lenovo x1, all targets may be affected
- [x] Is this a new feature
  - [ ] Verify user setup script works properly at first boot
  - [ ] Verify all desktop functionality works as before
  - [ ] Verify all targets build and run
  - [ ] Verify known errors/inconveniences as stated in 'Implementation Notes'
  - [ ] Verify `passwd` can be used to change user password

- [ ] If it is an improvement how does it impact existing functionality?

### CODE REVIEW

File: modules/microvm/virtualization/microvm/audiovm.nix [hunk 1]
```diff
@@ -72,6 +72,17 @@ let
             ] ++ lib.optional config.ghaf.development.debug.tools.enable pkgs.alsa-utils;
           };
 
+          users.users."proxy-user-audio" = {
+            isNormalUser = true;
+            uid = config.ghaf.users.accounts.loginuid;
```
Comment (brianmcgillion) [inline modules/microvm/virtualization/microvm/audiovm.nix:77]: why?

File: modules/microvm/virtualization/microvm/netvm.nix [hunk 2]
```diff
@@ -139,6 +139,15 @@ let
             ${config.ghaf.security.sshKeys.waypipeSshPublicKeyDir}.options = [ "ro" ];
           };
 
+          users.users."proxy-user-network" = {
+            isNormalUser = true;
+            createHome = false;
+            uid = config.ghaf.users.accounts.loginuid;
```
Comment (brianmcgillion) [inline modules/microvm/virtualization/microvm/netvm.nix:145]: again, why can it not be unique service account?

**Comment by UnikieHR** [comment]
image

File: modules/microvm/virtualization/microvm/audiovm.nix (hunk 1 above)
Comment (mbssrc) [reply modules/microvm/virtualization/microvm/audiovm.nix]: As discussed, dbus needs user account with same uid for authorization

File: modules/microvm/virtualization/microvm/netvm.nix (hunk 2 above)
Comment (mbssrc) [reply modules/microvm/virtualization/microvm/netvm.nix]: as above

**Comment by milva-unikie** [comment]
Tested on Lenovo-X1 (full re-installation with installer)
- I was able to create an user account, but can not login with it. Gives an error: `/run/current-system/sw/bin/ghaf-session: line 8: /tmp/session.labwc.log: Permission denied`
- Login with ghaf/ghaf worked, but after rebooting I was asked to re-create the user account again.

Am I doing something wrong?

**Review by brianmcgillion — APPROVED** [review]

**Comment by brianmcgillion** [comment]
> Tested on Lenovo-X1 (full re-installation with installer)
> 
> * I was able to create an user account, but can not login with it. Gives an error: `/run/current-system/sw/bin/ghaf-session: line 8: /tmp/session.labwc.log: Permission denied`
> * Login with ghaf/ghaf worked, but after rebooting I was asked to re-create the user account again.
> 
> Am I doing something wrong?

which commit are you using? i had the same issue on a slightly older commit from this morning. but the latest commit b8beea903ad7de73ae20d611e4bd813f15a1ff25 is working nicely. i can store the password and login.
 
other thing is the locale of the keyboard. could it change from password creation to first use?

**Comment by milva-unikie** [comment]
> > Tested on Lenovo-X1 (full re-installation with installer)
> > 
> > * I was able to create an user account, but can not login with it. Gives an error: `/run/current-system/sw/bin/ghaf-session: line 8: /tmp/session.labwc.log: Permission denied`
> > * Login with ghaf/ghaf worked, but after rebooting I was asked to re-create the user account again.
> > 
> > Am I doing something wrong?
> 
> which commit are you using? i had the same issue on a slightly older commit from this morning. but the latest commit [b8beea9](https://github.com/tiiuae/ghaf/commit/b8beea903ad7de73ae20d611e4bd813f15a1ff25) is working nicely. i can store the password and login.
> 
> other thing is the locale of the keyboard. could it change from password creation to first use?

I had the latest version. Re-installed and now the login also works with installed image. Maybe I did typo something the first time.

The current issue we are solving is how to get test-automation to work. Will continue with actual testing tomorrow morning.

File: packages/wifi-signal-strength/default.nix
Comment (kajusnau) [inline packages/wifi-signal-strength/default.nix:1]: Should double check, but this script is likely not used anywhere anymore and can be removed entirely.

Comment (mbssrc) [reply packages/wifi-signal-strength/default.nix:1]: removed

**Comment by mbssrc** [comment]
Changes in [79b4a8c](https://github.com/tiiuae/ghaf/pull/827/commits/79b4a8c4dd0d3e9d7fc5f074de6052a82d4b8033):
- change systemd dependency of login-user setup script to `greetd.service` to allow test automation, i.e., ssh to gui-vm with `ghaf` user before login user account is setup
- remove signal-strength script as it is unused

**Comment by milva-unikie** [comment]
**Tested on Lenovo-X1** (full re-installation)

Created my own user account, logged into it and used that to test.

Issues:
- Can not open images or pdf files from File Manager.
- Integrated camera does not work in business-vm.
- USB camera does not work in chrome-vm.
- Mounted USB sticks are not shown in File Manager, user can't access the files.
- Test-automation has many fails. This is already being investigated.

I have not had as much time to test as I would like, but to save time I wanted to add these comments already. I will continue testing with X1. With other targets we probably should wait until some fixes have been implemented.

**Comment by mbssrc** [comment]
## Updates with [b8a9aef](https://github.com/tiiuae/ghaf/pull/827/commits/b8a9aef153d2c4802158254d51575b8accc0b233:):

- [ ] Can not open images or pdf files from File Manager: Bug fixed.

- [ ] Integrated camera does not work in business-vm.
- [ ] USB camera does not work in chrome-vm.

     Seems unrelated to this patch - disabling apparmor will allow camera access in both. Was this working before?

- [ ] Mounted USB sticks are not shown in File Manager, user can't access the files.

     Code is fine, but default UID is incompatible. This causes devices with ext2/3/4 not being accessible.
     To increase usability, default UID for login user is now changed to 1000, which is the default UID on 
     most linux systems. 

- [ ] Support for test automation
  - [ ] New service to automatically setup test user. Procedure:
      1. Ssh to gui-vm
      2. Issue `(sudo) systemctl start setup-test-user.service`. This will create a new user `testuser` with password `testpw` as per [test-automation patch](https://github.com/tiiuae/ci-test-automation/pull/192)
  - [ ] Host service to remove current user setup. Note this currently does NOT delete data in appvms. Usage (on host): `systemctl start remove-users.service`

**Comment by johannarautanen** [comment]
**Checked with native NX and AGX.**

**Issues:**
- cannot open the application-menu from up-left. 
[just.txt](https://github.com/user-attachments/files/18146974/just.txt)

- cannot open the application-menu with Win-button either. With main native-builds, there is the old bug that you cannot open the application menu with WIN-keyboard button, so cannot say if this issues is because of that old bug or not.
- The screen selections in up-left failed, there is 1,2,3,4-options, only "1" is used no matter what other number you choose.
- when you select "sudo su" the terminal path is:
Screenshot from 2024-12-16 10-30-38


**Checked with cross compile AGX:**

**Issues:**
- cannot open the application-menu from up-left.
- cannot open the application-menu with Win-button either. With main croosscompile-builds, is possible to open the application menu with win-keyboard button
- The screen selections in up-left failed, there is 1,2,3,4-options, only "1" is used no matter what other number you choose.
- - when you select "sudo su" the terminal path is the same as with native builds: /var/empty


**Working with native and crosscompile:**
- the power menu-options works: Lock, suspend etc works and calendar comes visible when tap the day.
- with mouse right button can open the Terminal-menu and terminal itself

**Comment by milva-unikie** [comment]
**Tested on Lenovo-X1** (full re-installation)

Working:
- Test-automation mostly passes when [patch1](https://github.com/tiiuae/ci-test-automation/pull/192) and [patch2](https://github.com/tiiuae/ci-test-automation/pull/191) are applied. There is still a small issue with some of the performance tests that needs to be fixed in the test-automation side.
- Usb stick works.
- Opening files from File Manager works.


Smaller issues:
- Can still login to desktop with ghaf/ghaf. That desktop does not work properly and can be quite confusing for the users.
- Some apps (at least Audio Control & Bluetooth Settings) were in light theme after booting. At some point they switched to dark theme. I did not yet investigate closer why this happened.

Notes:
- Camera issues are also in main. Bug has been created.

File: modules/microvm/virtualization/microvm/microvm-host.nix [hunk 3]
```diff
@@ -90,23 +100,58 @@ in
...
+              echo "All ghaf login user data removed"
+            '';
+          };
+        in
+        mkIf config.ghaf.profiles.debug.enable {
+          description = "Remove ghaf login users";
+          enable = true;
+          path = [ userRemovalScript ];
+          unitConfig.ConditionPathExists = "/storagevm/gui-vm/var/lib/nixos/user.lock";
+          serviceConfig = {
```
Comment (kajusnau) [inline modules/microvm/virtualization/microvm/microvm-host.nix:142]: **StandardOutput** defaults to the value of **DefaultStandardOutput**, which itself defaults to **journal**.
Therefore, we can probably remove `StandardOutput = "journal"` and `StandardError = "journal"`, unless they are needed for clarity.

Ref.
https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html

**Comment by kajusnau** [comment]
> **Checked with native NX and AGX.**

> * The screen selections in up-left failed, there is 1,2,3,4-options, only "1" is used no matter what other number you choose.

This is not necessarily an indication of whether or not desktop switching is working. To confirm, testing should include opening any window in the active workspace/desktop, then switching desktops by any means and seeing if the window disappears and reappears as expected. If yes, then the issue is related to the workspace tracking file.

By default, the workspace tracking file is located in `~/.config/labwc/current-workspace`. If the file is not available or cannot be created, the described behavior is expected.
The file should be initially created by `labwc-autostart` script located in `modules/desktop/graphics/labwc.config.nix`.

**Comment by johannarautanen** [comment]
> > **Checked with native NX and AGX.**
> 
> > * The screen selections in up-left failed, there is 1,2,3,4-options, only "1" is used no matter what other number you choose.
> 
> This is not necessarily an indication of whether or not desktop switching is working. To confirm, testing should include opening any window in the active workspace/desktop, then switching desktops by any means and seeing if the window disappears and reappears as expected. If yes, then the issue is related to the workspace tracking file.
> 
> By default, the workspace tracking file is located in `~/.config/labwc/current-workspace`. If the file is not available or cannot be created, the described behavior is expected. The file should be initially created by `labwc-autostart` script located in `modules/desktop/graphics/labwc.config.nix`.
> 
> Edit: checking logs of ewwbar service might also be beneficial - `journalctl --user-unit ewwbar --no-pager --no-hostname`

When you tried to choose another desktop, the chosen desktop is not selected, stay still in "1" desktop:
Screenshot from 2024-12-16 12-08-25

**Comment by kajusnau** [comment]
> When you tried to choose another desktop, the chosen desktop is not selected, stay still in "1" desktop: Screenshot from 2024-12-16 12-08-25

Looks like home directory is not set correctly, it's defaulting to `/var/empty` instead. Hopefully this is helpful info to fix the bug for @mbssrc

**Comment by mbssrc** [comment]
## Update a66f5eb

- Desktop & admin home issue:
  - create home for admin user conditionally (AGX/NX)
  This should fix most/all issues on AGX. Tested and seems to work fine.

- Prevent admin login:
  - add "desktop" group; added for `admin` in AGX/NX, `loginuser` for laptop target
  - add dependency for `ghaf-session.target`, add `requires` attribute for ewwbar.service to prevent starting it and 
    any subsequent services if target fails

**Comment by johannarautanen** [comment]
**Checked with native and crosscompile NX and AGX.**

**Notes:**
- when you select in net-vm "sudo su" the terminal path is:
Screenshot from 2024-12-17 07-33-31

**Working:**
- application menu can be open from the up left
- application menu open also with win-keyboard button
- desktops can be changed successfully
- applications can be launched
- the power menu-options works: Lock, suspend etc works and calendar comes visible when tap the day.
- with mouse right button can open the Terminal-menu and terminal itself
- "sudo su" in host, terminal path now:
Screenshot from 2024-12-17 09-24-44


- automation tests: there is one issue, but that will be fixed in test-automation "setup" side. Other cases were ok.

**Review by vunnyso — APPROVED** [review]

**Comment by milva-unikie** [comment]
**Tested on Lenovo-X1** (full re-installation)

Working:
- Test-automation mostly passes when https://github.com/tiiuae/ci-test-automation/pull/192 and https://github.com/tiiuae/ci-test-automation/pull/191 are applied. Still some work needed in the test-automation side to get the rest of the tests to pass. 
- Usb stick works.
- Opening files from File Manager works.

Issues:
- Can still login to desktop with ghaf/ghaf. That desktop does not work properly and can be quite confusing for the users.
- If you login to ghaf, logout and then try to login to your own account, the login does not work. `/run/current-system/sw/bin/ghaf-session: line 8: /tmp/session.labwc.log: Permission denied ` After a reboot login works again.

I can create a bug about the issues if this pr gets merged with them.

**Comment by johannarautanen** [comment]
**Checked with Risc-V**

**Working:**
- able to boot up the Risc-V
- able to login to ghaf-host and check ghaf-and nixos-versions
- able to reboot via command line
- ethernet connection worked

No new issue found in manual testing.

**Review by kajusnau — APPROVED** [review]

**Pull request merged**